_posts/outbox-what-and-why.md

```diff
@@ -0,0 +1,155 @@
+5. Acknowledge the incoming message, deleting it from the queue.
+
+Not having an incoming message changes things a bit. We need a way to start a local transaction to store data and dispatch intentions. If we cannot use the incoming message to trigger the local transaction, we could use an ASP.Net middleware that begins the local transaction when the HTTP request starts. We can use the same middleware to decide when to commit the transaction, and that is when the HTTP request finishes. With that set, we can tweak the endpoint steps in the following way:
+
+1. When the HTTP request starts, start a local transaction.
+2. Handle the request and store any message dispatch intent.
+3. Commit the local transaction.
+4. Send messages we want to send and mark them as sent (outside the original local transaction).
+
+Now, what if we fail at step four? We're in trouble because we need something to hook on to retry. There is no incoming message. If step four fails, the HTTP request fails. Theoretically, we could retry the incoming HTTP request, but that's only an in-memory retry logic. What if step four "fails" because we never reach step four, and the web application crashes right after committing the local transaction?
+
+We could borrow some of the "outbox with an incoming message" logic and do the following:
+
+1. The middleware starts a local transaction when the HTTP request starts.
+2. Handle the request and store any message dispatch intent.
+3. Send a control message to ourselves.
+4. Commit the local transaction.
+5. Send messages we want to send and mark them as sent (outside the local transaction opened at step one).
+
+In step three, the endpoint sends a control message to itself. When received, the role of the control message is to check if step five succeeded in dispatching and marking messages as dispatched. If not, it'll deliver messages. At this point of the flow, we're in the same scenario of an "outbox with an incoming message." If the control message is faster than the local transaction commit phase, it'll find nothing, the regular message retry logic will kick in, and it'll eventually succeed.
+
+Now, let's look again at failure scenarios for other steps of the process:
+
+- Step one, two, or three fail: We do nothing; the HTTP request's endpoint allows the exception to bubble up to the HTTP client, for example, with an HTTP 500. It's a common synchronous request/response failing scenario. Since it happens inside a transaction, the endpoint rolls back every change, and the system remains consistent and will dispatch no messages.
+- Step four fails: Similarly to the previous scenario, if the endpoint fails in committing the local transaction, everything is left unchanged, and an error will bubble up to the client. However, the control message will arrive at the local queue. It'll find nothing because there were no changes. It'll retry a few times and eventually end up in the error queue, where it can be safely discarded.
```

Comment:
```suggestion
- Step four fails: Similar to the previous scenario, if the endpoint fails in committing the local transaction, everything is left unchanged, and an error will bubble up to the client. However, the control message will arrive at the local queue. It'll find nothing because there were no changes. It'll retry a few times and eventually end up in the error queue, where it can be safely discarded.
```